\documentclass[article, oneside]{memoir}
\usepackage{candor, verbatim, abstract}
\usepackage[charter]{mathdesign}
\renewcommand{\abstractname}{}
% Suppress memoir's fancy headings.
\pagestyle{plain}
\linespread{1.2}

\hyphenation{frame-shifts gen-bank da-ta-base hy-bri-di-za-tion}

\newcommand{\BWFtitle}[1]{A Computational Model for Translational
  Efficiency and Frameshifts in #1{Escherichia coli} Using a Genetic Signal
  Processing Approach}
\newcommand{\BWFauthors}{Hao Lian, Vivek Bhattacharya, and Daniel
  R. Vitek}

\usepackage[final, colorlinks=true, linkcolor=BWFBlue,
  citecolor=BWFGreen, urlcolor=BWFRed, pdftitle={\BWFtitle{}},
  pdfauthor={\BWFauthors}, pdfcreator={The Frameshift Kids},
  pdfstartview={FitH}]{hyperref}

\author{\BWFauthors}
\title{Abstract}

\begin{document}
\maketitle
In genetics, \ecoli\ is used as an expression system to commercially
produce proteins. However, sequence-dependent features, such as rare
codons and codon biases, affect translational efficiency. To tackle
this problem, we proposed a stochastic model to computationally
estimate translational efficiency and predict frameshifting, uniting
ideas from biological literature and developing two predictive
metrics. We ran our model on 4364 sequences from the \ecoli\ genome
and found over 90\% of them to have predicted high yields; moreover,
the model predicts ribosomal proteins to translate at even higher
rates---both results that concur with experimental evidence. We
investigated a set of eight sequences of bovine growth hormone, and
the model correctly determined the translational efficiency for seven.
We then examined variations of \prfB, a gene with a programmed
frameshift; the model grouped these sequences into two general
categories of high and low yield, consistent with experimental
results. Successful development of a computational model for
translational efficiency implicates itself in optimizing recombinant
protein yield in multiple fields, including commercial protein
synthesis.
\end{document}
