\documentclass[12pt, draft]{article}
\usepackage{candor, setspace}
\linespread 2
\numberwithin{equation}{section}

\begin{document}
\section{Introduction}

\section{Description of the Model}
\subsection{Free Energy}
\label{freeenergy}

Free energy in a cell arises from the hybridization between two
sequences of RNA and drives ribosomal translation~\cite{starmer}.
Weiss et al.~\cite{weiss} suggests the 16S tail of the ribosome hybridizes with its mRNA strand;
from this, we can calculate the free energy values for the interaction between the two strands of RNA.
Freier et al.~\cite{freier}, in turn, proposes a thermodynamic model for exactly this,
modeling the interaction in terms of \emph{doublets}, i.e. pairs of consecutive nucleotides.
Using this model, Freier numerically calculated the free energy available
for the hybridization of all RNA doublets permutations.

% Hao: I took out the setting energy values due to irrelevance.
From this, we can simulate hybridization between the associated tRNA, correspond to the number
of times the force can increment displacement from the current reading frame.
In their deterministic model, the force acts for \emph{exactly} this set number of cycles given a codon.

\subsection{Frameshifts}

% Where the following concept should be first mentioned, I'm not sure -- Vivek

First, we let a displacement of $x = 0$ correspond to the zero reading frame and increments of
\emph{two} to represent a one-nucleotide change. For example, $x =2$ represents a +2 frameshift.
Ponnala~\cite{lalit:embs} prove that both $x = 0$ and $x = 2$ are fixed (stable) points in displacement in his model.

% Figure prfB:lalitdisp will be prfB on Lalit's model.  Figure prfb:polar is the polar plot for prfB. Obviously.  --Vivek
% 26 or 25? What counting system are we using? More details needed, but stick with 25 for now. --Hao

A sudden jump from approximately $x = 0$ to $x = 2$ is then the first indication of a $+1$ frameshift.
In essence, this jump suggests that the ribosome jumps one entire base pair in the mRNA sequence.
Figure~\ref{prfB:lalitdisp} shows the displacement plot as per this deterministic model for \prfB, 
a gene known to have a programmed $+1$ frameshift at codon 25.

In conjunction with this characteristic plot, a $+1$ frameshift also shows a 240\degree
rotation of phase angle from the species angle, consistent with the creation of displacement data from the phasors.
Intuitively, a frameshift is only sustained if the free energy signal aligns itself with the sudden jump in displacement.
Since the free energy signal has a period of one codon~\cite{lalit:mechanics}, for a $+1$ frameshift the free energy signal
must undergo a phase shift of one-third of an entire period (Figure~\ref{prfb:polar}).

\subsection{A Stochastic Model for Displacement}

% Find an example of a gene with a screwed-up displacement plot under Lalit's old model. --Vivek

The gene \prfB\ exhibits a characteristic frameshift under the deterministic model: The displacement jumps to $x=2$.
Certain genes, however, demonstrate ambiguous behavior near $x = \pm 1$.
In a deterministic model, the fixed displacement plots implies
we cannot clearly interpret whether this behavior indicates a tendency for a stable frameshift.
In practice, ribosomal translations are not deterministic. Due to the presence of
noise within the cell environment, we instead chose to model translation stochastically, adding
randomness to the choice of reading frame a ribosome executes at each codon given the free energy signal.
As such, we introduce probability into the model.

We propose that at each cycle in elongation, the ribosome must make a decision: stay in the current reading frame,
move to the $+1$ reading frame,
or move to the $-1$ reading frame.  We can further subdivide this choice into the individual wait-cycles.
At each wait cycle, the ribosome chooses from the above possibilities or proceeds to another wait-cycle without making a decision.

Let $abcd$ be a sequence of four nucleotides, with $abc$ in the
current reading frame and $bcd$ in a +1 reading frame.  Let $x$ be the
displacement of the current wait cycle of the ribosome.  As the
incremental displacement approaches +1, the probability of choosing
codon $bcd$ should increase and the probability of choosing codon
$abc$ should decrease.  We modeled this behavior using even powers of
cosine and sine functions for $abc$ and $bcd$.  We
define $\omega$ as the weight that is directly proportional to
the probability.

\begin{equation}
  \omega_{abc} = \cos^{10}{\frac{x\pi}{4}} \text{ and } \omega_{bcd} = \sin^{10}{\frac{x\pi}{4}}.
\end{equation}

Now we define a value $n_{abc}$ based on normalizing $N_{abc}$, the
number of wait cycles, itself inversely proportional to the
tRNA availability of codon $abc$.
We assume the probability of frameshifting after
after $N_{abc}$ wait cycles is $\frac{1}{2}$.  Let $P$ be the
probability of moving on to the next codon and staying in current reading
frame.  Then we should have $1-\left(1-P\right)^{N_{abc}} =
\frac{1}{2}$.  Because $P \propto \omega_{abc}$, we define $n_{abc}$ to be the constant of
proportionality. That is, $n_{abc} = \omega_{abc} / P$.  Coupled
with that $\omega_{abc} \le 1$, we have $n \le \sqrt[N]{2}/(\sqrt[N]{2} - 1).$
where $n = n_{abc}$ and $N = N_{abc}$. Mathematically, the probability of choosing a codon is then
\begin{align}
  \prod_N \left(1-\frac{\omega}{n}\right) \text{ where } \omega = \cos^{10}{\frac{x\pi}{4}}.
\end{align}
Here, $n$ represents the wait time for that codon and $N$ is the ordinal of the wait cycle. (For example,
$N=1$ on the first wait cycle, $N=2$ on the second, and so forth.)

\section{Applications}

\subsection{Bovine Growth Hormone}
Besides the frameshifting sequence \prfB, our model also shows a significant correlation in data with biological
bovine growth hormone (bGH) data, sequenced in \ecoli~\cite{schoner:bgh}. For example, our model purports a relation
between the deviation from the zero reading frame---for non-frameshifters---and protein yield, postulating
an association between ribosomal efficiency and its reading frame stability. Schoner et al.~\cite{schoner:bgh}
modified beginning codons of a sequenced bGH sequence to produce a series of alternates,
two of which produced significantly higher yield in experiments. Those two, pcZ101 and pcZ105, procured significantly
lower deviation than pcZ104, pcZ108, and other bGH sequences in our model, which agrees Schoner's findings.
A computational model's accurate prediction of a sequence's yield greatly benefits microbiology,
aiding in the measure of a protein's translational efficiency especially when experiments prove too difficult
or costly. Therefore, bGH is a cornerstone of our computational model's scientific importance.

\subsection{rpoS RNA Polymerase Protein}
Our model can also calculate the sample probability that a gene sequence will run correctly without any mistake frameshifts
effectively due to the computational and fast nature of running it.
At high sample sizes it readily approximates the actual probability and therefore efficiency, which we have shown with sequences such as
\prfB\ and ribosomal proteins. In addition, research~\cite{rare:yield}
suggests we can remove trouble spots where mistake frameshifts occur by replacing those codons and their
earlier neighbors with synonymous codons\footnote{Two codons are synonymous if and only if they produce the same amino acid.}.
Therefore, we wrote a locally optimal (greedy) algorithm that finds trouble spots and replaces them with random sequences
of synonymous codons. The algorithm runs those permutations through our model and chooses the
sequence with the lowest likelihood of frameshifting incorrectly. (Hence, locally optimizing.) One ribosomal protein with
low yield is \rpoS, referred to us via Susan Gottesman and Dr. Anne-Marie Stomp.
Running \rpoS\ through our algorithm, we found 33 codon changes that produced a higher yield, soon hopefully biologically verified.
If it proves that our modified \rpoS\ does indeed express higher levels
of RNA polymerase, then our model represents a breakthrough in optimizing gene sequences for an untold number of  applications,
especially commercial production. As with bGH, our model has the capability of replacing a tedious biological experimentation
with a computationally fast process.

\footnotesize
\begin{singlespace}
  \bibliography{wizards}
\end{singlespace}
\end{document}
