\documentclass[12pt]{article}
\usepackage{candor}
\linespread 2
\begin{document}

%SWC = Sexy White Chicks.  Not Siemens-Westinghouse Competition.  Are you so much of a nerd that you would rather win a research competition than have a hot girl around?  (Vivek: the answer is yes.)

\section{Introduction}

\section{Description of the Model}

\subsection{Free Energy}
\label{freeenergy}

% Find correct citations for the next paragraph.  NOW.

Free energy in a cell, which arises from the hybridization between two
sequences of RNA, drives ribosomal translation~\cite{starmer}.

Research suggests the 16S tail of the ribosome hybridizes with its  mRNA~\cite{weiss};
from this, we can  calculate the free energy values for the interaction between the two strands of RNA.
Someone et al.~\cite{freier} propose a thermodynamic model for exactly this,
modelling the interaction in terms of \emph{doublets}, i.e. pairs of consecutive nucleotides.
Using this model, Freier et al.~\cite{freier} numerically calculated the free energy available
upon hybridization for all permutations of RNA doublets.

We use Freier et al.'s values to calculate a free energy value at each nucleotide.
First, the 13-base 16S tail of \ecoli\ hybridizes with the first 13 bases of a sequence, which
for a gene is a 12-base leader sequence and the first base of the start codon, to determine
the free energy associated with the first nucleotide.
Then, shifting one base pair at a time, we calculate the free energy for the entire sequence.
Since any free energy value greater than zero represents binding that would only take place if energy were added,
we set such values to zero.

% We need to introduce the concept of displacement somewhere; I'm not sure where, though -- Vivek

\subsection{A Deterministic Model for Displacement}

Ponnala et al.~\cite{lalit:mechanics} assumes a sinusoidal model for free energy, whence one can calculate
the magnitude and phase of ribosomal translation using arctan given the free energy signals once the memory model
also proposed by Ponnala et al. stores those values in a phasor memory structure,
where memory here is biologically equivalent to the displacement from the reading frame of the ribosome performing the translation
and the registers represent a snapshot of the free energy values at a given codon during translation. The magnitude and phase
of the ribosomal translation here represents the phasor Ponnala uses to structure the register contents and are used to generate
polar plots.

Ponnala et al.~\cite{lalit:embs} represents the cummulative phasor at codon $k$ as $\vec{V}_k = Me^{j\theta_k}$.
Differentiating this vector, we arrive at vector $\vec{D}_k$ for instantaneous energy, interpreting this
vector as a force that acts on the ribosome, keeping the mRNA in a given reading frame.

The length of time that the force acts on the ribosome depends up the codon's tRNA availability at the A-site during translation.
Ponnala use a deterministic model to represent this: For each codon, a number
of ``wait cycles," dependent upon the rarity of the associated tRNA, correspond to the number
of times the force is allowed to increment displacement from the current reading frame.
In their deterministic model, the force acts for \emph{exactly} this set number of cycles, given a certain codon in a certain gene sequence.

\subsection{A Stochastic Model for Displacement}

However, in practice it is idealistic to assume that all sequences, especially those that frameshift, will be read in exactly the same manner every single time that a given mRNA strand passes through the ribosome.  In order to account for this essentially random 

\subsection{Frameshifts}

% Where the following concept should be first mentioned, I'm not sure -- Vivek

As mentioned earlier, a displacement of $x = 0$ corresponds to translation along the 0 reading frame,
while a displacement of $x = 2$ represents translation along the $+1$ reading frame.
Ponnala et al.~\cite{lalit:embs} prove that both $x = 0$ and $x = 2$ are stable points in displacement.

% Figure prfB:lalitdisp will be prfB on Lalit's model.  Figure prfb:polar is the polar plot for prfB. Obviously.  --Vivek

A sudden jump from approximately $x = 0$ to $x = 2$ is the first indication of a $+1$ frameshift; 
in essence, this jump suggests that the ribosome jumps one entire base pair in the mRNA sequence.
Figure~\ref{prfB:lalitdisp} shows the displacement plot---by this deterministic model---for \prfB, 
a gene known to have a programmed $+1$ frameshift at codon 26.

In conjunction with this characteristic plot, however, a $+1$ frameshift also shows a 240\degree
rotation of phase angle from the species angle.
Intuitively, a frameshift would only be sustained if the free energy signal aligns itself with the sudden jump in displacement.
Since a the free energy signal has a period of one codon, for a $+1$ frameshift---a one-nucelotide jump---the free energy signal
must undergo a phase shift of one-third of an entire period.  Figure~\ref{prfb:polar} indicates this shift in \prfB.

\subsection{A Stochastic Model for Displacement}

% Find an example of a gene with a screwed-up displacement plot under Lalit's old model. -- Vivek

The gene \prfB\ shows a very characteristic frameshift under the deterministic model; the displacement jumps almost exactly to $x=2$.
Certain genes, however, demonstrate ambiguous behavior near $x = 1$.
Coupled with the limitation that a deterministic model will output a fixed graph for a given sequence, 
it is difficult to interpret whether this behavior indicate a tendency for a stable frameshift.
As such, we introduce probability into the model.

We propose that at each cycle in elongation, the ribosome must make a decision: stay in the current reading frame, move to the $+1$ reading frame,
or move to the $-1$ reading frame.  We can further subdivide this choice into the individual wait-cycles: 
At each wait cycle, the ribosome can choose one of the above possibilities or proceed to another wait-cycle without making a decision.

The probabilities of choosing

$$\Pr\left(abc\right) = 1 - \prod_{i=1}^j \left(1-\frac{\omega_{abc}}{n_{abc}}\right)$$

$$\omega_{abc} = \cos^{10}{\frac{x\pi}{4}}$$

\section{Applications}

\subsection{Bovine Growth Hormone}
Besides the frameshifting sequence \prfB, our model also shows a significant correlation in data with biological
bovine growth hormone data, sequenced previously in \ecoli~\cite{schoner:bgh}. For example, our model purports a relation
between the deviation from the zero reading frame---for non-frameshifters anyway---and protein yield, ultimately postulating
a relation between ribosomal efficiency and its ability to maintain its reading frame without wobbling. Schoner et al.
modified beginning codons of an initial bovine growth hormone sequence to produce a series of alternate bGH sequences,
two of which produced significantly higher yield than their compatriots. Those two, dubbed pcZ101 and pcZ105, produced significantly
lower deviation than pcZ104, pcZ108, and other bGH sequences, thus agreeing with the biological data from Schoner et al.
A model's accurate prediction of yield given a sequence would greatly benefit the work of biologists across the world,
aiding in verifying the yields of a large sequence of proteins especially when experiments on all of them proves too difficult
or costly. Therefore, bGH is an important cornerstone of our computational model's impact on biology.

\subsection{rpoS}

\bibliography{wizards}
\end{document}