\documentclass[12pt]{article}
\usepackage{candor, setspace}
\usepackage[verbose]{wrapfig}
\hyphenation{frame-shifts}
\newcommand{\BWFtitle}{Computationally
  Modeling Ribosomal Translation and Programmed Frameshifts in}
\newcommand{\BWFauthors}{Hao Lian, Vivek Bhattacharya, and Daniel
  R. Vitek}

\usepackage[final, colorlinks=true, linkcolor=BWFBlue,
  citecolor=BWFGreen, urlcolor=BWFRed, pdftitle={\BWFtitle\ Escherichia
    coli}, pdfauthor={\BWFauthors},
  pdfsubject={Bioinformatics, Genetics, and Biology}, pdfcreator={The
    Frameshift Kids}, pdfkeywords={bioinformatics,genetics,biology,
    frameshifts,perl,matlab,ecoli,prfB,rpoS,bGH,ecoli}, pdfstartview={FitH},
  backref]{hyperref}

\linespread 2
\numberwithin{equation}{section}

\author{\sc{\BWFauthors}}
\date{{\sc \today}}
\title{\bf{\BWFtitle\ \emph{Escherichia coli}}}

\begin{document}
% Title, table of contents, and abstract start things off.
\pagenumbering{roman}
\begin{singlespace}
  \maketitle
  \tableofcontents
\end{singlespace}

\clearpage
\begin{abstract}\begin{normalsize}
  In \emph{\BWFtitle\ Escherichia coli}, we discuss a stochastic model for ribosomal displacement relative to reading 
  frame based on forces that arise from changes in free energy present in hybridization between the 
  16S rRNA tail and nucleotides upstream from the A-site.  We present both our investigation and 
  potential applications for this model, including optimization of economically significant gene
  sequences such as bovine growth hormone, non-experimental prediction of translational efficiency for man-made 
  proteins, adjustment of calculated tRNA availability values, and modelling of frameshifts in \prfB\ in \ecoli.
  We present this investigation as a series of various inquiries made 
  in subjects ranging throughout the field of molecular biology.  We also applied our stochastic 
  model to create a short, 16-codon sequence that is not known to produce a frameshift in nature 
  in order to test our model's predictive power for synthesized mRNA sequences, and designed an 
  experiment to verify this frameshift.  We expect results in October 2007.
\end{normalsize}\end{abstract}
\clearpage
\pagenumbering{arabic}

\section{Introduction}
Prior research on translational efficiency, especially sequence-dependent features, has shown that 
translational efficiency can be influenced by such factors as codon bias~\cite{gustafsson04} and 
rare codon usage~\cite{kane95}~\cite{ikemura}.  These sequence-dependent factors even apply to 
expression of recombinant proteins~\cite{sorensen05}.  In addition, secondary-structure problems 
associated with the folding of the sequence can decrease translational efficiency~\cite{kozak05}. 
However, the specific connections between translational efficiency and these factors are not fully 
understood, and eliminating troublesome factors does not always correlate to increased efficiency. 
[Dr. Stomp, we need references here.]  We have developed an additional tool that can help to 
eliminate problems caused by codon bias and codon usage, which will certainly be useful.  

The recent work of~\citet{lalit:embs} focused on creating a deterministic model of frameshifting based on 
hybridization between the 16S rRNA tail and upstream mRNA nucleotides.  The periodicity of this signal 
discussed in [Ponnala-JBSB.pdf --need to cite] suggests that force arising from the free energy in such hybridizations acts as a correcting mechanism, helping to ensure that possible overshooting or 
undershooting of the next in-frame codon does not occur.  In this manner, early errors in reading frame tend
not to be propagated throughout the entire sequence.  Some limitations of the deterministic model are
discussed in section 2.4.

Because the ribosome's displacement from the zero reading frame is virtually always nonzero, and because the 
free energy acts as a correcting mechanism, we reasoned that we could create a new metric for a sequence: 
the total deviation from the intended reading frame.  This metric, because it measures how far the ribosome 
is from optimal functionality, should then roughly correlate to translational efficiency.  Throughout this 
paper, we use our new metric to optimize tRNA availabilities in order to distinguish between high-efficiency
sequences and low-efficency sequences.  We also use our new metric to optimize a gene sequence known to be 
translationally regulated for maximum efficiency, and will apply our new metric to ribosomal proteins and 
the entire \ecoli genome to see if the model is robust over many different types of genes.

% Insert introduction here.
% It has been done. -Daniel
\section{Computational Methods}
\subsection{Free Energy}
\label{freeenergy}

Free energy in a cell arises from the hybridization between two
sequences of RNA and drives ribosomal translation~\cite{starmer}.
\citet{weiss88} suggest the 16S tail of the ribosome hybridizes with its mRNA strand;
from this, we can calculate the free energy values for the interaction between the two strands of RNA.
\citet{freier} propose a thermodynamic model for exactly this,
modeling the interaction in terms of \emph{doublets}, which are pairs of consecutive nucleotides.
Using this model, Freier calculated the free energy available
for the hybridization of permutations of RNA doublets.

\subsection{The Deterministic Model}
\citet{lalit:mechanics} assume a sinusoidal model for
free energy, whence we can project free energy onto magnitude and
phase through a memory model. Ponnala simulates free
energy in a memory structure that can store three values (registers)
that he later can represent as a phasor, a concept from physics. Then,
we can visualize free energy from polar plots and deduce frameshifts,
which occur along defined boundaries~\citet{lalit:mechanics} on the polar plot for a species.
 
\citet{lalit:embs} then represent the cumulative phasor
at codon $k$ as $\bvec{V} = Me^{i\theta}$ where $i$ is the imaginary
constant, allowing us to calculate the magnitude and phase at codon
$k$ through simple trigonometry. Magnitude and phase are then modeled
as a polar plot, one of the model outputs. Differentiating the energy vector $\bvec{V}$
with respect to distance along the mRNA strand, we
arrive at vector $\bvec{D}$ and interpret it as
a force that acts on the ribosome, keeping the mRNA in
a given reading frame.
 
The length of time that the force acts on the ribosome depends upon
the codon's tRNA availability at the A-site during translation as well as the its shape [Dr. Stomp, references please!].
\citeauthor{lalit:mechanics} use a deterministic model to represent this: For each codon,
a number of ``wait cycles," dependent upon the rarity of the
associated tRNA, correspond to the number of times the force can
increment displacement from the current reading frame.  In the
deterministic model, the force acts for \emph{exactly} this set number
of cycles given a codon. We can then simulate hybridization between the
16S ribosomal subunit and a given mRNA strand: First, the 13-base 16S
tail of \ecoli\ hybridizes with the first 13 bases of a sequence,
which is a 12-base leader sequence and the first base of the start
codon \textsc{aug}, to determine the free energy value of the first acid.
Then, shifting one base pair at a time, we calculate the free energy
for the entire sequence according to \citet{starmer}.

\subsection{Frameshifts}
\label{section:frameshifts}

\begin{cfigure}
  \caption{Plots of~\prfB}
  \label{prfB:detplots}
  \subfloat[Deterministic displacement]{
    \label{prfB:deterministic:sub}
    \includegraphics[width=0.5\textwidth]{prfB/deterministic}
  }
  \subfloat[Polar plot]{
    \label{prfB:polar:sub}
    \includegraphics[width=0.4\textwidth]{prfB/polar}
  }
\end{cfigure}

First, we let a displacement of $x = 0$ correspond to the zero reading
frame and increments of \emph{two} to represent a one-nucleotide
change. For example, $x =2$ represents the +1 frame.
\citet{lalit:embs} prove that both $x = 0$ and $x = 2$ are fixed
(stable) points in displacement in their model, as expected.

A jump from approximately $x = 0$ to $x = 2$ {\emph{over the span of just one base pair}} is then the first
indication of a $+1$ frameshift; it suggests the ribosome skips one
entire base pair in the mRNA sequence.  \autoref{prfB:deterministic:sub}
shows the displacement plot per this deterministic model for \prfB, a
unique gene with a programmed $+1$ frameshift at codon 25. In
conjunction with this characteristic plot, a $+1$ frameshift also
displays an equally characteristic clockwise 120\degree\ phase angle
rotation from the species angle, the average phase of the free energy
signals of a number of verified \ecoli\ genes that stay in frame \cite{lalit:mechanics}.  Intuitively, a
frameshift sustains if the free energy signal aligns with the sudden
jump in displacement.  Since the free energy signal has a period of
one codon~\cite{lalit:mechanics}, a $+1$ frameshift the free energy
signal must undergo a phase shift of one-third of an entire period
(\autoref{prfB:polar:sub}).

\subsection{A Stochastic Model for Displacement}
\label{stochastic}

As mentioned, the gene \prfB\ exhibits a programmed frameshift under
the deterministic model: The displacement jumps to $x=2$.  Certain
other genes, however, demonstrate equivocal, ambiguous behavior near
$x = \pm 1$~\cite{lalit:embs}.  In a deterministic model, we lack the
information and sensitivity needed to discern clearly a tendency for a
stable frameshift. Even worse, the model may not show these unstable
behaviors at all. In practice then, ribosomal translations are not
deterministic. Due to the presence of noise within the cell
environment, any deterministic model ultimately imperfectly accounts
for translation. Therefore, we next explored stochastically modeling
ribosome translation with a sinusoidal probability paradigm that
follows.

We propose that at each cycle in elongation, the ribosome must make a
decision: stay in the current reading frame, move to the $+1$ reading
frame, or move to the $-1$ reading frame.  We can further subdivide
this choice into the individual wait-cycles.  At each wait cycle, the
ribosome chooses from the above possibilities or proceeds to another
wait-cycle without making a decision.  In addition, we assert the 
number of wait cycles is inversely proportional to the tRNA availability of 
the codon in the current reading frame, as rarer codons should force the 
ribosome to wait longer for a correct transfer. In turn, we calculate
tRNA availability based on existing research suggesting an association
between it and codon frequency~\cite{ikemura}.

Let $abcd$ be a sequence of four nucleotides, with $abc$ in the
current reading frame and $bcd$ in a +1 reading frame.  Let $x$ be the
displacement of the current wait cycle of the ribosome.  As the
incremental displacement approaches +1, the probability of choosing
codon $bcd$ should increase and the probability of choosing codon
$abc$ should decrease.  We model this behavior using even powers of
cosine and sine functions for $abc$ and $bcd$.  We
define $\omega$ as the \emph{weight} that is directly proportional to
the probability.
\begin{equation}
  \omega_{abc} = \cos^{10}{\frac{x\pi}{4}} \text{ and } \omega_{bcd} =
  \sin^{10}{\frac{x\pi}{4}}.
  \footnote{For the purposes of this model, the cosine and sine
    functions are taken to the tenth power, but in future studies,
    this parameter, which must be an even integer, can change.}
\end{equation}
If the ribosome lies completely in the 0 frame, then the probability
of staying in the 0 frame should be 1.  Consequently, if the ribosome
lies fully in the +1 frame ($x=2$), the probability of going to the 0
frame should be 0. Therefore, these functions have a period of two
base pairs ($x=4$).

We define $N_{abc}$ to be the number of wait cycles 
allocated to the given codon in the current frame.
Suppose we are at a particular wait cycle on codon $abc$ and
abbreviate $N = N_{abc}$ as per above.
In absence of further research, we assume the probability
of frameshifting is 1/2.  Let $P$ be the instantaneous probability of
moving on to the next codon and staying in the current reading frame.
Then we should have $1 - \left(1-P\right)^{N} = \frac{1}{2}$.
Given that $P \propto \omega$, let $n = n_{abc}$ be the constant of
proportionality; thus $n = \omega / P$.  We have
$\omega \le 1$, so hence $n \le \sqrt[N]{2}/(\sqrt[N]{2} - 1)$
Mathematically, the
probability of choosing the codon at a given wait cycle in the zero reading frame is just

\begin{equation}
  \prod_{i=1}^K \left(1-\frac{\omega_i}{n}\right) \text{ where }
  \omega_i = \cos^{10}{\frac{x_i\pi}{4}}.\footnote{The sinusoidal
    component of $\omega_i$ is not always cosine and depends upon the
    frame choice in question.}
\end{equation}

Here, $n$ represents $n_{abc}$ for the current codon and $K$ is the
ordinal of the current wait cycle. For example, $K=1$ on the first
wait cycle, $K=2$ on the second, and so forth.  Importantly, $\omega$
is dependent on displacement, which increments after every wait cycle.

We also have $\displaystyle\lim_{N\rightarrow\infty} n = N/\ln{2}$, indicating
that the probability of choosing a codon at a given cycle is directly proportional
to its tRNA availability, another result that coincides with reason and intuition.


\subsection{Consequences of the Stochastic Model}

The stochastic model has introduced a new concept into the model proposed by
\citet{lalit:mechanics}: Now, the ribosome can ``choose the wrong codon," in 
essence accepting the tRNA of a codon out-of-frame, either due to a high availability
of the tRNA or due to a more applicable shape.  It is imperative to note
that this phenomenon is not equivalent to frameshifting.  In a programmed
frameshift, the force pushes the ribosome to an unstable point---one at which
the ribosome must move to the +1 reading frame quickly in order to maintain
equilibrium.  Thus, programmed frameshifts take place over the span of simply
one codon; the graph never approaches $x=2$ over a number of codons.

An incorrect codon choice occurs when displacement approaches $x = \pm 1$.
At this point, the +1 frame is in view, along with the 0 frame.  As the probability
equations suggest, the ribosome is quite liable to choose the select the codon
in the +1 frame by accident.  This action, unlike a frameshift, is a result
of a digression that takes place over a number of codons---a digression that
occurs slowly.  As the ribosome nears $x = 1$, the values for both the sine 
and cosine functions drop, thus increasing the probability of indecision, that is,
increasing the time required for translation.

Choosing the wrong codon is a purely stochastic phenomenon; only through our new
model can we actually track the decisions of the ribosome throughout the course
of translation.  Due to a general biological conceptualization of the physics involved,
as described above, we believe that the stochastic model has the potential to 
measure translational efficiency.


\section{Analysis}
In order to test our computational model, we ran a number of
experiments to analyze sequences present in the \ecoli\ genome.

\subsection{Measures}
\label{section:metrics}

As this model is stochastic, multiple runs (the sample size) of the same sequence must be analyzed.
As such, we propose two metrics for analysis of a number of different runs 
simultaneously. 

\subsubsection{Error-Free Rate}
\label{section:efr}
When studying a
sequence with a programmed frameshift, \emph{error-free rate} measures the percentage of runs 
during which the ribosome chooses the correct codon
at \emph{every} juncture.  For a +1 frameshift, for example, the ribosome must
choose the +1 frame at the frameshift codon and stay in the 0 frame before
and the +1 frame after in order for the run to be a success.

\subsubsection{Displacement Deviation}
\label{section:deviation}

We define \emph{displacement deviation} to be
\begin{equation}
    d = \sqrt{\frac{\sum_i \left(x_i - \beta_i\right)^2}{N}},
\end{equation}
where $\beta_i$ is the predicted reading frame at codon $i$, $x$ is
the displacement at codon $i$, and $N$ is the total number of codons
as a measure of the deviation of the sequence from the expected
reading frame.  Usually $\beta_i = 0$ unless a programmed frameshift
exists as it does for \prfB.  For example, for \prfB\ $\beta_i = 2$
for all $i \geq 25$ because \prfB\ frameshifts at codon 25
\textsc{uga} and the model represents a frameshift with +2
displacement per \autoref{section:frameshifts}.

\subsection{\prfB\ and Related Sequences}
As discussed, \prfB\ codes for protein release factor B in \ecoli.
Notably, microbiologists agree that \prfB\ has a programmed frameshift
in the 25$^\textrm{th}$ codon~\cite{weiss87}.  We thus test the
gene \prfB\ to determine whether our model can predict this
frameshift.  We downloaded the nucleotide sequence for \prfB\ from
NCBI's Genbank database.  [Dr. Stomp, do we need a citation for Genbank?]

\citet{weiss87} performed a number of biological experiments on
\prfB\ to test how mutations in the sequence would affect the rate of
frameshifting.  They present a total of 35 sequences in their paper,
along with relative measures of translational efficiency.  We ran all
these through our model and correlated them with their error-free rates.
We hypothesize that genes found to frameshift at high rates by
\citeauthor{weiss87} should also show high error-free rates under
our model.

\subsection{The \ecoli\ Genome and Ribosomal Proteins}
We now aim to test the capabilities of our model to predict translational
efficiency other than frameshift rate.  We hypothesize that displacement
deviation provides a suitable metric for translational efficiency:  A lower
deviation would correspond to a more efficient sequence.  

[Dr. Stomp, are there any references in literature to the relationship
  between ribosome displacement instability and protein efficiency?]
We base this hypothesis on the fact that as the displacement moves
away from the horizontal axis, the ribosome has a greater probability
of picking the codon in the +1 frame by chance, in effect
\emph{probabilistically} producing junk. That is, while a single run
may biologically produce a working primary structure of the protein,
the gene in question ultimately is less efficient than a synonymous gene
with lower aggregate probabilities be this aggregate measure the
error-free rate (\autoref{section:efr}) or displacement deviation
(\autoref{section:deviation}). In addition, a greater deviation from
the zero axis inherently implies a longer time for translation, since
the force must act on the ribosome for an extended period of time to
make the displacement relatively large.  A high value thus
indicates a propensity for a high time for translation, reducing
translational efficiency, again contributing to the aggregate
probability of translational failure.

To this end, we start with a simple fact:
Microbiologists agree that translation in \ecoli\ is an efficient process;
most regulation is transcriptional.  As such, the majority of the genes 
in the genome should exhibit low displacement deviations.  Moreover,
ribosomal proteins are known to be abundant in the cell, and thus
are hypothesized to be translated at especially efficient rates.  We
furthermore predict that ribosomal proteins will have deviations lower
on average than those for the other \ecoli\ genes.

We run the model on a set of 4364 genes collected from \ecoli.  The
nucleotide sequences were obtained from Ecogene.

\subsection{Bovine Growth Hormone}
We finally test our model on a set of sequences known to be translationally 
regulated.  \citet{schoner:bgh} developed a number of sequences that
code for variations of bovine growth hormone. \citeauthor{schoner:bgh}
note that these sequences are, in fact, translationally regulated.
\citeauthor{schoner:bgh} present a set of eight sequences, four of which
are expressed at high levels, and four or which have negligible expression.
We predict that the sequences that are highly expressed in
\ecoli\ should have lower displacement deviations.

\subsection{Parameters}
\label{section:parameters}
For all computational experiments in this report, we use a species
angle of $\theta_{\rm{sp}} = -30\degree$ and an initial displacement of 0.1,
in accordance with \citet{lalit:embs}.  Later in this report, we explore
the effects of changing the species angle and initial displacement on the
error-free rate of \prfB.

A parameter that is much more difficult to estimate
is tRNA availability value (TAV) vectors.
\citeauthor{lalit:embs} base the tRNA availability values on codon usage, 
surveying a set of genes from \ecoli.
Although this assumption has biological basis~\cite{ikemura}, 
it is possible that the true values for these numbers may differ 
due to sampling error or the choice of genes chosen.
Taking a different track, we chose to calibrate TAV vectors from
existing biological data, designing a genetic algorithm to calibrate
these values based on the bovine growth hormone (bGH) sequences
while still remaining numerically close to the values determined by \citeauthor{lalit:embs}.

We optimize the separation between the displacement deviations of the 
high-yield sequences and the other ones.  
First, we generate a list of randomly modified TAV vectors at which
point we calculate the ratio of the displacement deviation from
displacement (\autoref{section:deviation}) for the four high-yield 
sequences to the other bGH proteins. From there, we sort the 
modified TAV according to this ratio and discarding the least 
optimal half of the tRNA vectors. We then choose two of the 
remaining vectors, with the chance of each vector being proportional
 to its rank.  We then take a weighted average of the two vectors 
 and spawn a new vector.  After creating the next generation according 
 to a constant ``gene pool'' size, we delete the previous generation 
 and repeat. After a fixed number of generations, the algorithm 
 terminates and returns the optimal TAV vector.

This algorithm did not alter the TAV vectors to a significant degree.
In fact, the average change to each value in the vector was merely
15.99\%.  Yet, the newly generated vector returns much more optimal
results than values from \citeauthor{lalit:embs}.  The new values
are used throughout the remainder of the paper to produce the 
values and graphs.

\section{Results}
\subsection{\prfB}

\begin{cfigure}
  \caption{Plots of \prfB\ in a stochastic model}
  \label{prfB:stochplots}
  \subfloat[Displacement Plot]{
    \label{prfB:disp:sub}
    \includegraphics[width=0.4\textwidth]{prfB/disp}
  }
  \subfloat[Sensitivity plot]{
    \label{prfB:sens:sub}
    \includegraphics[width=0.5\textwidth]{prfB/sensitivity}
  }
\end{cfigure}

The gene \prfB, as mentioned, is known
to have a programmed frameshift at the 25$^{\textrm{th}}$ codon.
\autoref{prfB:disp:sub} shows a displacement plot for
\prfB, again with a distinctive jump at codon 25.
\footnote{Note that the polar plot is the the same as \autoref{prfB:polar:sub}.
The new model does not alter the polar plot or the free energy calculations.}

Notably, the displacement plot does not reach $x=2$ over the span of
one codon, as the old model predicted.  Rather, due to randomness, the
ribosome chooses the codon in the $+1$ frame before actually reaching
a displacement of exactly 2.  The propensity to approach $x=2$,
however, concurs with biological evidence indicating the ribosome
stays in frame after the frameshift.  [Dr. Stomp, we need references
  here.]
  
\autoref{prfB:sens:sub} shows the error-free rate of \prfB\ as a function
of species angle and initial displacement.  This plot is used mainly to
demonstrate robustness, as will be detailed in the discussion.

\begin{wrapfigure}{R}{0.5\textwidth}
  \caption{Comparison of biological yield and error-free rate, 500 iterations}
  \label{weissboxplot}
  \includegraphics[width=0.5\textwidth]{histograms/weissbox}
\end{wrapfigure}

We repeated \citeauthor{weiss87}'s experiments computationally
using the stochastic model, noticing a marked contrast between those
with \bgals\ activity over 1650 whole-cell units, low in
relationship with \prfB\ that had an activity of 6600, and those
without. We distinguish between these as high and low yield, respectively.
With some reservations, the data supports our prediction that high yield
and low yield sequences could more-or-less be distinguished via
error-free rate as we found from our results.

We must note that \citeauthor{weiss87} did not maintain the amino acid structure
while substituting codons.  Changes in the primary structure is of interest in
analysis of sequences.

\subsection{\ecoli\ Genes}
\begin{cfigure}
  \caption{Investigating a large sample of \ecoli\ genes}
  \subfloat[Displacement deviations]{
    \label{ecoli:hist}
    \includegraphics[width=0.4\textwidth]{histograms/everything}
  }
  \quad
  \subfloat[Comparison to ribosomal proteins]{
    \label{ribosomal:comp}
    \includegraphics[width=0.5\textwidth]{histograms/ribosomal}
  }
\end{cfigure}

\autoref{ecoli:hist}\footnote{We truncate the histogram at three, excluding
  the outliers and less than 1\% of the sample.} is a histogram of the
deviation yields of 4364 genes of \ecoli, over 80\% of the entire
genome.  As predicted, 93.45\% of the genes that we ran laid in the 0
to 1 interval, agreeing with our theory of natural efficiency.  The
average displacement deviation for these \ecoli\ genes is 0.4425, in a
run for 500 iterations per gene.

\subsection{Ribosomal Proteins}
\label{section:riboproteins}
To compound this finding, biologists also agree that ribosomal
proteins offer especially high translational efficiencies, since the
cell must produce them in such large quantities. [Dr. Stomp, we need
  references here.] The displacement deviation from $x=0$ for ribosomal proteins
is on average 0.2708 in comparison to the average of 0.4425 for our
large sample of \ecoli\ genes, which is significantly higher with a $t$-value of
0.0109 when performing a two-sample $t$ test on the means.

\subsection{Bovine Growth Hormone}
\label{section:bgh}

\begin{cfigure}
  \caption{Displacement plot for bGH}
  \label{bgh:disp}
  \includegraphics[width=0.5\textwidth]{bgh/all}
\end{cfigure}

We investigated the concept of deviation yield as a measure of biological
yield by studying bovine growth hormone (bGH), a protein commonly used
in agriculture.
Research~\cite{schoner:bgh} at the time attempted to produce bGH
in \ecoli\ in large amounts.

\input{bgh/deviations}

\citet{schoner:bgh}, primarily modifying the initial codons of an
initial bovine growth hormone sequence, found sequences pcZ101,
pcZ105, pcZ112, and pcZ115, to have particularly high protein yield
and to be efficient in comparison to the six other sequences. In
modeling displacement, we found these four sequences  to have the least
displacement deviation from $x = 0$
(\autoref{bgh:deviation}). \autoref{bgh:disp} shows the displacement
plots of all the bGH sequences on the same set of axes. Although not
immediately obvious, the four sequences do indeed indicate lower
deviation. Despite running the genetic algorithm, pcZ108 remains an
outlier because \citeauthor{schoner:bgh} reported low protein
efficiency whereas our model reports low deviation, incongruous with
previously established correlations.

Once again, it must be noted that like Weiss, \citeauthor{schoner:bgh}
did not maintain the amino acid structure while altering his sequences.
The sequence pcZ108 has a different amino acid structure than the other
ones; the lower translational efficiency may be attributed to the interaction
of the primary structure with the ribosome and not to any signal encoded
in the mRNA.

\subsection{Optimizing Translational Efficiency}

If displacement deviation is in fact a potential metric for biological yield,
a method for optimizing deviation would be of use of biologists.  As such,
we wrote an algorithm to minimize deviation for given sequences.
We tested this algorithm computationally on \rpoS, a gene known to
be translationally regulated.  Biological experiments will be
conducted soon.

% I'm colluding deviation and probability yield here, but it's for a
% good purpose. --Hao.

[Dr. Stomp, we need the notes that you mentioned.]
\citet{rpoS:process} indicate that \rpoS, which codes for an RNA
polymerase sigma factor, contains sections of rare codons that disrupt
ribosomal translation. Our computation model agrees, showing
relatively high deviation from $x = 0$ in comparison to other
ribosomal proteins (\autoref{section:riboproteins}). In mass
production of such a polymerase sigma factor, biologists replace
sequences of codons with synonymous counterparts known to reduce noise
and error. However, with a computation model, the process is much
faster and with this performance we constructed a randomized and greedy
algorithmic search for such replacements. We first find an
early trouble spot\footnote{Specifically, places with mistake
  frameshifts in our model, which correlate to rare codons per above
  discussion.} of four codons, randomly replacing that sequence with
synonymous four codons, and running our model against those
permutations to obtain a locally optimal sequence at that place. We
then repeat for all trouble spots the first, terminating when we have
locally optimized the last one. With this algorithm, we reduced our
standard deviation metric for the \rpoS\ displacement plot from 0.168
to 0.117 on sample size of 1000 with a replacement of 33 codons. The
initial correlation between deviation and biological expression
provides strong anecdotal evidence that, with future biological
experimentation, our algorithm has indeed increased protein yield
bypassing a potentially slow biological experimentation process.

\subsection{An Artificial Frameshifter}
\begin{cfigure}
  \caption{Artificial linker sequence sensitivity plot}
  \label{linker:sens}
  \includegraphics[scale=0.25]{linker/sensitivity}
\end{cfigure}

\begin{cfigure}
  \caption{Artificial linker sequence with a 12-base leader sequence
    in brackets}
  \label{linker}
  \begin{verbatim}
    [aga aau cag acc] aug gag gcu ggc acc agg ggg uac agu  u  aag caa acg
  \end{verbatim}
\end{cfigure}

In order to verify the model's ability to predict frameshifts, we
artificially created a sixteen-codon sequence (\autoref{linker})
designed to frameshift.  The sequence is designed to frameshift into
the \textsc{aag} frame after crossing the sole uracil.  Figure~\ref{linker:sens}
shows the error-free rate of the linker sequence as a function of species
angle and initial displacement, in order to demonstrate robustness,
similar to the \prfB\ sensitivity plot (\autoref{prfB:sens:sub}).

A BLAST search finds about thirty matches of our sequence within areas of
bacterial genomes but none from \ecoli. To test the frameshifting capacity of
this linker sequence, we initiate a biological experiment.

We create a bifunctional fusion protein containing beta-galactosidase (\bgals) 
and \xylE. \bgals\ levels correlate to
levels of nitrophenol, which has a yellow color, while \xylE\ expresses
an enzyme that cleaves catechol.  These two proteins are joined by
the proposed linker sequence.  Research indicates that even in this
fused protein, both \bgals\ and \xylE\ should retain their functionalities.
As such, if the frameshift were to occur, one should observe both 
high levels of nitrophenol and low levels of catechol.

We passed on these initial parameters to a graduate student with
microbiological expertise who has constructed a plasmid of an
ampicillin gene, the \emph{lacZ} operon, our linker sequence, and the
catechol enzyme gene \xylE\ in order to measure the protein
efficiency of our linker sequence in addition to other control
groups. We expect results in October 2007.

\section{Discussion}
First, we note that our model is quite robust: Minor changes
in $\theta_{\rm{sp}}$ and initial displacement do not affect
frameshifting significantly.
\autoref{prfB:sens:sub} illustrates the error-free rate as
a function of species angle and initial displacement. As noted,
the frameshift holds over quite a wide range. Though these initial
parameters are necessary (\autoref{section:parameters}), exact values
are not
crucial to the true interplay within the model between free energy and
translational efficiency.

In addition, our model agrees with a great pool biological
evidence. In our model, ribosomal proteins express more highly than
genes in general as per \citet{rpoS:process}
(\autoref{section:riboproteins}).  Our correlation in running bovine
growth hormone again indicate this crucial relationship between
translational efficiency~\cite{schoner:bgh} and the metrics of our
model. With continual refinement, our model can then can significantly
increase the speed at which geneticists and biologists can obtain
valuable information when synthesizing commercially or medicinally
useful compounds without working through potentially slow biological
experiments especially in the commercial value of bGH in this
instance.  These concrete evidence suggest that our model
accurately predicts protein synthesis efficiency with its metrics
(\autoref{section:metrics}) and those metrics have a strong
association with the metrics of protein production in biological experiments.

A motif throughout the results is that outliers---certain sequences by
Weiss and pcZ108, for example---differ in primary structure from
sequences known to translate at high rates. Although we designed our
model to concentrate on the signal analysis of the mRNA, there also
exists biochemical interactions between nascent polypeptides with and
their ribosomal or cellular environment.  As such, we cannot
distinguish between errors in translation due to encoding in the
signal and the actual protein structures. In future research, we will
collect more data on the process by which synonymous codons can
improve translational efficiency. We aim also to calibrate more finely
tRNA values with a data set beyond bovine growth hormone, which can
allow for larger deviation from given values~\cite{lalit:embs} than
the small margin of error we used. The underlying theme of future
research is then to congeal together more biological data and model
results from the vast pool of research with which our model has not
yet come into contact.

In addition, \citet{weiss87} conducted a set of experiments in which
they changed the codon before the frameshift site at {\sc{guu}},
dropping the yield to less than 60 whole-cell units. In contrast,
our model predicts very high error-free rates at an average of 92.8\%
for all six sequences. Again, our model does not fully correlate with
all biological evidence in the field and \citeauthor{weiss87}'s
experiments remain a topic of scrutiny as we continually refine our
model in future research.

% Frameshift discussion

The model certainly provides considerable predictive power over a broad spectrum
of biological topics, exemplified by the investigations in this report.  Yet,
as we reconcile assumptions of the model with biological
experimentation, we find wrinkles future research needs to address.
For example, choosing the wrong codon suggests the ribosome always produces
long proteins, producing some dysfunctional proteins. However, such is not the case
[Dr. Stomp, we need some clarification or citations here]. We also
assumed deviation from the baseline entails a longer wait time to
model the dissociation of the ribosome from the mRNA. With future
research, we aim to refine this assumption with more rigor and
biological evidence as we consider the biological implications of the model in addition
to its predictive power.

\phantomsection
\addcontentsline{toc}{section}{References}
\begin{singlespace} \bibliography{wizards} \end{singlespace}
\end{document}

% This is for ispell. Do not delete. --Hao
% LocalWords:  embs abcd abc bcd sp pcZ riboproteins disp
