#summary All you need to know
#labels Featured,documentation,manual

== Introduction ==

(For a theoretical treatment of this project, [http://theframeshiftkids.googlecode.com/files/Siemens%20Paper.pdf please read our research paper].) Biologists can tell how well _Escherichia coli_ translates a certain mRNA strand with various experimental methods. Is there a way to replace that process or parts of it with a computer?

=== Terminology ===

A _gene file_ refers to a file in the FASTA format. However, our model simply ignores the FASTA header line that describes the gene. As a result, you can also input a file that is purely the sequence of bases that make up a gene. For simplicity's sake, the de facto file extension we've adopted for this relaxed FASTA version is just `.txt`, which is how we name all the genes we publish and all the genes to which we shall refer in this manual.

== Structure ==

Our model is divided into two sections: Matlab and Perl. The entirety of the Perl section lies in the `pearls` folder. Each file has (or should have) complete POD documentation. That is, if you type the program name at the command line without any arguments, the help text generated by `Pod::Usage` should show up. Try it out with `scan_brightly.pl`. On the Matlab side, the documentation is written in comment form for each file. Each Matlab file corresponds to one major, documented (hopefully) function. To access it, use `help` from Matlab's interactive environment. Try it out with `help unity.m`.

The Matlab files within the `unities` folder are thematically united by the purpose of data collection. Each file collects a specific statistic (for example, deviation from the zero reading frame) usually over a number of samples and many, many genes. As a result, many `unities` can take a directory _or_ a gene file as an argument.

The files within the `God 2.0` and the `thrushbaby` directories are part of now-defunct algorithms. They are further described in the "Algorithms" section.  The files within `util` are utility functions used by some Matlab files. You might find them useful in constructing new tools based on our model. They're friendly folks with open hearts and beautiful faces.

== Pearls ==

=== Getting started with free energies ===

Free as in C memory, not free as in you burnt your thighs with our coffee and now we have a settlement. The primary Perl program exposed to users is the `scan_brightly.pl` file, which is an interface to our `Kidnap` Perl module. The command is simple: `scan_brightly.pl prfB.txt` and a list of free energies at each codon spills into your command prompt. Usually you will never have to type this because the list is useless without the Matlab portion of our model. However, let this not mislead you to believe that it's not important. Our entire model rests upon these free energies. They are our cornerstone and, most importantly, our friends.

For a more in-depth treatment of all the pearls, see [ManualsPearls the Pearls section of the manual].