\documentclass[12pt, draft]{article}
\usepackage{candor, setspace}
\linespread 2
\numberwithin{equation}{section}

\begin{document}
\section{Introduction}

\section{Description of the Model}
\subsection{Free Energy}
\label{freeenergy}

Free energy in a cell arises from the hybridization between two
sequences of RNA and drives ribosomal translation~\cite{starmer}.
Weiss et al.~\cite{weiss} suggests the 16S tail of the ribosome hybridizes with its mRNA strand;
from this, we can calculate the free energy values for the interaction between the two strands of RNA.
Freier et al.~\cite{freier}, in turn, proposes a thermodynamic model for exactly this,
modeling the interaction in terms of \emph{doublets}, i.e. pairs of consecutive nucleotides.
Using this model, Freier numerically calculated the free energy available
for the hybridization of all RNA doublets permutations.

% Hao: I took out the setting energy values due to irrelevance.
From this, we can simulate hybridization between the associated tRNA, correspond to the number
of times the force can increment displacement from the current reading frame.
In their deterministic model, the force acts for \emph{exactly} this set number of cycles given a codon.

\subsection{Frameshifts}

% Where the following concept should be first mentioned, I'm not sure -- Vivek

First, we let a displacement of $x = 0$ correspond to the zero reading frame and increments of
\emph{two} to represent a one-nucleotide change. For example, $x =2$ represents a +1 frameshift.
Ponnala~\cite{lalit:embs} prove that both $x = 0$ and $x = 2$ are fixed (stable) points in displacement in his model.

A sudden jump from approximately $x = 0$ to $x = 2$ is then the first indication of a $+1$ frameshift.
In essence, this jump suggests that the ribosome jumps one entire base pair in the mRNA sequence.
Figure~\ref{prfB:lalitdisp} shows the displacement plot as per this deterministic model for \prfB, 
a gene known to have a programmed $+1$ frameshift at codon 25.

In conjunction with this characteristic plot, a $+1$ frameshift also shows a 240\degree
rotation of phase angle from the species angle, consistent with the creation of displacement data from the phasors.
Intuitively, a frameshift is only sustained if the free energy signal aligns itself with the sudden jump in displacement.
Since the free energy signal has a period of one codon~\cite{lalit:mechanics}, for a $+1$ frameshift the free energy signal
must undergo a phase shift of one-third of an entire period (Figure~\ref{prfb:polar}).

\subsection{A Stochastic Model for Displacement}

% Find an example of a gene with a screwed-up displacement plot under Lalit's old model. --Vivek

The gene \prfB\ exhibits a characteristic frameshift under the deterministic model: The displacement jumps to $x=2$.
Certain genes, however, demonstrate ambiguous behavior near $x = \pm 1$.
In a deterministic model, the fixed displacement plots implies
we cannot clearly interpret whether this behavior indicates a tendency for a stable frameshift.
In practice, ribosomal translations are not deterministic. Due to the presence of
noise within the cell environment, we instead chose to model translation stochastically, adding
randomness to the choice of reading frame a ribosome executes at each codon given the free energy signal.
As such, we introduce probability into the model.

We propose that at each cycle in elongation, the ribosome must make a decision: stay in the current reading frame,
move to the $+1$ reading frame,
or move to the $-1$ reading frame.  We can further subdivide this choice into the individual wait-cycles.
At each wait cycle, the ribosome chooses from the above possibilities or proceeds to another wait-cycle without making a decision.

Let $abcd$ be a sequence of four nucleotides, with $abc$ in the
current reading frame and $bcd$ in a +1 reading frame.  Let $x$ be the
displacement of the current wait cycle of the ribosome.  As the
incremental displacement approaches +1, the probability of choosing
codon $bcd$ should increase and the probability of choosing codon
$abc$ should decrease.  We modeled this behavior using even powers of
cosine and sine functions for $abc$ and $bcd$.  We
define $\omega$ as the weight that is directly proportional to
the probability.

\begin{equation}
  \omega_{abc} = \cos^{10}{\frac{x\pi}{4}} \text{ and } \omega_{bcd} = \sin^{10}{\frac{x\pi}{4}}.
\end{equation}

Now we define a value $n_{abc}$ based on normalizing $N_{abc}$, the
number of wait cycles, itself inversely proportional to the
tRNA availability of codon $abc$.
We want to wait long enough that the probability of assume the probability of frameshifting after
after $N_{abc}$ wait cycles is $\frac{1}{2}$.  Let $P$ be the
probability of moving on to the next codon and staying in current reading
frame.  Then we should have $1-\left(1-P\right)^{N_{abc}} =
\frac{1}{2}$.  Because $P \propto \omega_{abc}$, we define $n_{abc}$ to be the constant of
proportionality. That is, $n_{abc} = \omega_{abc} / P$.  Coupled
with that $\omega_{abc} \le 1$, we have $n \le \sqrt[N]{2}/(\sqrt[N]{2} - 1).$
where $n = n_{abc}$ and $N = N_{abc}$. Mathematically, the probability of choosing a codon is then
\begin{align}
  \prod_{i=1}^N \left(1-\frac{\omega_i}{n}\right) \text{ where } \omega = \cos^{10}{\frac{x\pi}{4}}.
\end{align}
Here, $n$ represents the wait time for that codon and $N$ is the ordinal of the wait cycle. (For example,
$N=1$ on the first wait cycle, $N=2$ on the second, and so forth.)
It is important to note that $\omega$ is dependent on displacement, 
which increments after every wait cycle.

\section{Applications}

\subsection{rpoS RNA Polymerase Protein}
Our model can also calculate the sample probability that a gene sequence will run correctly without any mistake frameshifts
effectively due to the computational and fast nature of running it.
At high sample sizes it readily approximates the actual probability and therefore efficiency, which we have shown with sequences such as
\prfB\ and ribosomal proteins. In addition, research~\cite{rare:yield}
suggests we can remove trouble spots where mistake frameshifts occur by replacing those codons and their
earlier neighbors with synonymous codons\footnote{Two codons are synonymous if and only if they produce the same amino acid.}.
Therefore, we wrote a locally optimal (greedy) algorithm that finds trouble spots and replaces them with random sequences
of synonymous codons. The algorithm runs those permutations through our model and chooses the
sequence with the lowest likelihood of frameshifting incorrectly. (Hence, locally optimizing.) One ribosomal protein with
low yield is \rpoS, referred to us via Susan Gottesman and Dr. Anne-Marie Stomp.
Running \rpoS\ through our algorithm, we found 33 codon changes that produced a higher yield, soon hopefully biologically verified.
If it proves that our modified \rpoS\ does indeed express higher levels
of RNA polymerase, then our model represents a breakthrough in optimizing gene sequences for an untold number of  applications,
especially commercial production. As with bGH, our model has the capability of replacing a tedious biological experimentation
with a computationally fast process.

\subsection{God}
Because our model is so dependent on tRNA availability, we decided to investigate what would happen if we changed tRNA availability.  However, we wanted to change the availabilities of various codons in ways that would help differentiate between high-yield sequences and low-yield sequences.  Because studies have already been done on the translational efficiency of the eight recombinant BGH proteins, we decided to try to calibrate our model using these sequences.  To this end, we developed a genetic algorithm described below.

Our algorithm starts by generating a list of randomly modified tRNA availability vectors.  At this point, we calculate $\mu_{low}$, the mean of the deviation yields of the six BGH sequences with lowest yields, and $\mu_{high}$, the mean of the deviation yields of the two BGH sequences with highest yields, namely PCZ101 and PCZ105\footnote{In Schoner's experiments, a signifcant difference was found between PCZ101 and PCZ105 compared to the other six sequences}.  We then sort the modified tRNA availabilities by the ratio $\mu_{high}/\mu_{low}$.  After discarding the worst half of the tRNA vectors, we choose two of the remaining vectors, with the chance of each vector being proportional to its rank.  We then take a weighted average of the two vectors and use this new vector as the basis for our next generation of random modifications.  After suitable improvements, the algorithm terminates.

\section{Applications: Bovine Growth Hormone}
One section of modern industry where genetics has proven useful is
agriculture where synthesized compounds keep the machinery of
production running day in and day out. For example, recent
research~\cite{schoner:bgh} have attempted to amplify the DNA sequence
that codes for bovine growth hormone in cows and attempt to produce it
in \ecoli. However, the process for modifying a DNA sequence is slow
and arduous in a biological experiment environment especially in the
measure of protein yield for each newly modified sequence. However,
with the help of our computation model, we can easily determine the
relative yields for each sequence quickly and efficiently if such a
correlation between the metrics our model outputs and biological
research exists. Indeed it does.

Schoner~\cite{schoner:bgh}, primarily modifying the initial codons of an initial
bovine growth hormone sequence, found two sequences pcZ101 and pcZ105,
to have particularly high protein yield and efficient with respect to
the six other sequences. Our model agrees. In modeling displacement,
we found pcZ101 and pcZ105 to have the least reading frame deviation
from $x = 0$~\ref{bgh:deviation}, which in our model~\cite{lalit:mechanics} indicates
higher yield as increasing deviation from the correct reading frame
produces larger error within the ribosome during translation.

\begin{table}[tbp]
\begin{center}
    \begin{tabular}{lc}
        \toprule
        \textbf{Sequence} & $\mathbf{\sigma}$\\
        \midrule
        pcZ101 & 0.76719\\
        pcZ105 & 0.76710\\
        \midrule
        pcZ100 & 0.80020\\
        pcZ104 & 0.79729\\
        pcZ108 & 0.79241\\
        pcZ110 & 0.79760\\
        pcZ112 & 0.79419\\
        pcZ115 & 0.79794\\
        \bottomrule
    \end{tabular}
    \caption{Deviation for bGH Sequences with Sample Size~200}
    \label{bgh:deviation}
\end{center}
\end{table}

In addition, in examining the polar plots, we noticed that during
early translation the phase of translation for pcZ101 and 105 tended toward the species
angle -30 degrees before diverging toward approximately 15 degrees. As
it stands, our model, as with displacement, indicates higher yield when the polar plot
indicates less deviation from the correct (zero) reading frame, as it
does here with the species angle~\cite{lalit:mechanics}. Therefore our
model is consistent with research~\cite{bgh:initiation} that indicates
the early codons have a dramatic impact on the ultimate efficiency of
ribosomal translation.

These results from bovine growth hormone that correlation with a broad
spectrum of known ribosomal behavior indicate our computational model
can significantly increase the speed at which geneticists and
biologists can obtain valuable information when synthesizing
commercially or medicinally useful compounds without laboriously
working through biological experiments.

\footnotesize
\begin{singlespace}
  \bibliography{wizards}
\end{singlespace}
\end{document}
